File: GRAPH THEORY/lab1/lab1_c++/lab1.cpp
#include <iostream>
#include <unordered_map>
#include <vector>
#include <fstream>
#include <cstdlib>
#include <ctime>
#include <set>

// Directed Graph class
class DirectedGraph {
private:
    std::unordered_map<int, std::unordered_map<int, int>> adjList; // adjacency list with costs (outbound)
    std::unordered_map<int, std::set<int>> inEdges; // inbound edges for efficiency
    int vertexCount = 0;

public:
    // Add a vertex
    void addVertex(int v) {
        if (adjList.find(v) == adjList.end()) {
            adjList[v] = {};
            inEdges[v] = {};
            vertexCount++;
        }
    }

    // Add an edge with a cost
    void addEdge(int src, int dest, int cost) {
        addVertex(src);
        addVertex(dest);
        if (adjList[src].find(dest) == adjList[src].end()) {
            adjList[src][dest] = cost;
            inEdges[dest].insert(src);
        } else {
            adjList[src][dest] = cost; // Update cost if edge exists
        }
    }

    // Remove an edge
    void removeEdge(int src, int dest) {
        if (hasEdge(src, dest)) {
            adjList[src].erase(dest);
            inEdges[dest].erase(src);
        }
    }

    // Remove a vertex and all connected edges
    void removeVertex(int v) {
        if (adjList.find(v) != adjList.end()) {
            // Remove outgoing edges
            for (const auto& edge : adjList[v]) {
                inEdges[edge.first].erase(v);
            }
            adjList.erase(v);
            // Remove incoming edges
            for (const auto& src : inEdges[v]) {
                adjList[src].erase(v);
            }
            inEdges.erase(v);
            vertexCount--;
        }
    }

    // Get the number of vertices
    int getVertexCount() const {
        return vertexCount;
    }

    // Get the number of edges
    int getEdgeCount() const {
        int count = 0;
        for (const auto& pair : adjList) {
            count += pair.second.size();
        }
        return count;
    }

    // Check if an edge exists
    bool hasEdge(int src, int dest) const {
        return adjList.find(src) != adjList.end() && adjList.at(src).find(dest) != adjList.at(src).end();
    }

    // Get in-degree of a vertex
    int getInDegree(int v) const {
        return inEdges.find(v) != inEdges.end() ? inEdges.at(v).size() : 0;
    }

    // Get out-degree of a vertex
    int getOutDegree(int v) const {
        return adjList.find(v) != adjList.end() ? adjList.at(v).size() : 0;
    }

    // Retrieve the cost of an edge
    int getEdgeCost(int src, int dest) const {
        if (hasEdge(src, dest)) {
            return adjList.at(src).at(dest);
        }
        return -1; // Edge not found
    }

    // Modify the cost of an edge
    void setEdgeCost(int src, int dest, int cost) {
        if (hasEdge(src, dest)) {
            adjList[src][dest] = cost;
        }
    }

    // Print all vertices
    void printVertices() const {
        std::cout << "Vertices: ";
        for (const auto& pair : adjList) {
            std::cout << pair.first << " ";
        }
        std::cout << "\n";
    }

    // Print all edges
    void printEdges() const {
        std::cout << "Edges:\n";
        for (const auto& pair : adjList) {
            for (const auto& edge : pair.second) {
                std::cout << pair.first << " -> " << edge.first << " (cost: " << edge.second << ")\n";
            }
        }
    }

    // Print outbound edges of a vertex (NEW)
    void printOutboundEdges(int v) const {
        std::cout << "Outbound edges from " << v << ":\n";
        if (adjList.find(v) != adjList.end()) {
            for (const auto& edge : adjList.at(v)) {
                std::cout << v << " -> " << edge.first << " (cost: " << edge.second << ")\n";
            }
        } else {
            std::cout << "No outbound edges.\n";
        }
    }

    // Print inbound edges of a vertex (NEW)
    void printInboundEdges(int v) const {
        std::cout << "Inbound edges to " << v << ":\n";
        if (inEdges.find(v) != inEdges.end() && !inEdges.at(v).empty()) {
            for (const auto& src : inEdges.at(v)) {
                std::cout << src << " -> " << v << " (cost: " << getEdgeCost(src, v) << ")\n";
            }
        } else {
            std::cout << "No inbound edges.\n";
        }
    }

    // Read graph from a file
    void readFromFile(const std::string& filename) {
        std::ifstream file(filename);
        if (!file) {
            std::cerr << "Error opening file!\n";
            return;
        }
        int n, m, src, dest, cost;
        file >> n >> m;
        adjList.clear();
        inEdges.clear();
        vertexCount = 0;
        for (int i = 0; i < m; ++i) {
            file >> src >> dest >> cost;
            addEdge(src, dest, cost);
        }
        file.close();
    }

    // Write graph to a file
    void writeToFile(const std::string& filename) const {
        std::ofstream file(filename);
        if (!file) {
            std::cerr << "Error writing to file!\n";
            return;
        }
        file << getVertexCount() << " " << getEdgeCount() << "\n";
        for (const auto& pair : adjList) {
            for (const auto& edge : pair.second) {
                file << pair.first << " " << edge.first << " " << edge.second << "\n";
            }
        }
        file.close();
    }

    // Generate a random graph
    void generateRandomGraph(int vertices, int edges) {
        adjList.clear();
        inEdges.clear();
        vertexCount = 0;
        for (int i = 0; i < vertices; ++i)
            addVertex(i);
        srand(static_cast<unsigned>(time(nullptr)));
        int count = 0;
        int maxEdges = vertices * (vertices - 1); // Max possible edges in a directed graph
        if (edges > maxEdges) {
            std::cout << "Too many edges! Max possible for " << vertices << " vertices is " << maxEdges << "\n";
            edges = maxEdges;
        }
        while (count < edges) {
            int src = rand() % vertices;
            int dest = rand() % vertices;
            if (src != dest && !hasEdge(src, dest)) {
                int cost = rand() % 100 + 1;
                addEdge(src, dest, cost);
                count++;
            }
        }
    }

    // Generate two random graphs and save them (NEW)
    void generateTwoRandomGraphs() {
        DirectedGraph g1, g2;

        // First graph: 7 vertices, 20 edges
        g1.generateRandomGraph(7, 20);
        g1.writeToFile("random_graph1.txt");
        std::cout << "First random graph (7 vertices, 20 edges) generated and saved as random_graph1.txt\n";

        // Second graph: 6 vertices, 40 edges (will be capped at 30 due to max edges)
        g2.generateRandomGraph(6, 40); // Max edges for 6 vertices is 6 * 5 = 30
        g2.writeToFile("random_graph2.txt");
        std::cout << "Second random graph (6 vertices, up to 40 edges) generated and saved as random_graph2.txt\n";
    }
};

void printMenu() {
    std::cout << "\n======== Directed Graph Menu ========\n";
    std::cout << "1. Add Vertex\n";
    std::cout << "2. Add Edge\n";
    std::cout << "3. Remove Vertex\n";
    std::cout << "4. Remove Edge\n";
    std::cout << "5. Get Vertex Count\n";
    std::cout << "6. Get Edge Count\n";
    std::cout << "7. Check Edge Existence\n";
    std::cout << "8. Get In-degree of Vertex\n";
    std::cout << "9. Get Out-degree of Vertex\n";
    std::cout << "10. Get/Set Edge Cost\n";
    std::cout << "11. Print Vertices and Edges\n";
    std::cout << "12. Generate Random Graph\n";
    std::cout << "13. Read Graph from File\n";
    std::cout << "14. Write Graph to File\n";
    std::cout << "15. Print Outbound Edges\n";      // NEW
    std::cout << "16. Print Inbound Edges\n";       // NEW
    std::cout << "17. Generate Two Random Graphs\n"; // NEW
    std::cout << "0. Exit\n";
    std::cout << "======================================\n";
    std::cout << "Enter your choice: ";
}

int main() {
    DirectedGraph g;
    int choice;

    do {
        printMenu();
        std::cin >> choice;

        if (choice == 1) {
            int v;
            std::cout << "Enter vertex (integer): ";
            std::cin >> v;
            g.addVertex(v);
            std::cout << "Vertex " << v << " added.\n";
        }
        else if (choice == 2) {
            int src, dest, cost;
            std::cout << "Enter source, destination and cost: ";
            std::cin >> src >> dest >> cost;
            g.addEdge(src, dest, cost);
            std::cout << "Edge added from " << src << " to " << dest << " with cost " << cost << ".\n";
        }
        else if (choice == 3) {
            int v;
            std::cout << "Enter vertex to remove: ";
            std::cin >> v;
            g.removeVertex(v);
            std::cout << "Vertex " << v << " removed.\n";
        }
        else if (choice == 4) {
            int src, dest;
            std::cout << "Enter source and destination of the edge to remove: ";
            std::cin >> src >> dest;
            g.removeEdge(src, dest);
            std::cout << "Edge from " << src << " to " << dest << " removed.\n";
        }
        else if (choice == 5) {
            std::cout << "Total vertices: " << g.getVertexCount() << "\n";
        }
        else if (choice == 6) {
            std::cout << "Total edges: " << g.getEdgeCount() << "\n";
        }
        else if (choice == 7) {
            int src, dest;
            std::cout << "Enter source and destination: ";
            std::cin >> src >> dest;
            std::cout << "Edge from " << src << " to " << dest << (g.hasEdge(src, dest) ? " exists.\n" : " does not exist.\n");
        }
        else if (choice == 8) {
            int v;
            std::cout << "Enter vertex: ";
            std::cin >> v;
            std::cout << "In-degree of vertex " << v << ": " << g.getInDegree(v) << "\n";
        }
        else if (choice == 9) {
            int v;
            std::cout << "Enter vertex: ";
            std::cin >> v;
            std::cout << "Out-degree of vertex " << v << ": " << g.getOutDegree(v) << "\n";
        }
        else if (choice == 10) {
            int subChoice;
            std::cout << "1. Get Edge Cost\n2. Set Edge Cost\nEnter choice: ";
            std::cin >> subChoice;
            if (subChoice == 1) {
                int src, dest;
                std::cout << "Enter source and destination: ";
                std::cin >> src >> dest;
                int cost = g.getEdgeCost(src, dest);
                if (cost != -1)
                    std::cout << "Cost of edge (" << src << ", " << dest << ") is " << cost << ".\n";
                else
                    std::cout << "Edge does not exist.\n";
            }
            else if (subChoice == 2) {
                int src, dest, cost;
                std::cout << "Enter source, destination and new cost: ";
                std::cin >> src >> dest >> cost;
                g.setEdgeCost(src, dest, cost);
                std::cout << "Edge cost updated.\n";
            }
        }
        else if (choice == 11) {
            g.printVertices();
            g.printEdges();
        }
        else if (choice == 12) {
            int vertices, edges;
            std::cout << "Enter number of vertices and edges for random graph: ";
            std::cin >> vertices >> edges;
            g.generateRandomGraph(vertices, edges);
            std::cout << "Random graph generated.\n";
        }
        else if (choice == 13) {
            std::string filename;
            std::cout << "Enter filename to read graph from: ";
            std::cin >> filename;
            g.readFromFile(filename);
            std::cout << "Graph loaded from file.\n";
        }
        else if (choice == 14) {
            std::string filename;
            std::cout << "Enter filename to write graph to: ";
            std::cin >> filename;
            g.writeToFile(filename);
            std::cout << "Graph written to file.\n";
        }
        else if (choice == 15) { // NEW: Print Outbound Edges
            int v;
            std::cout << "Enter vertex: ";
            std::cin >> v;
            g.printOutboundEdges(v);
        }
        else if (choice == 16) { // NEW: Print Inbound Edges
            int v;
            std::cout << "Enter vertex: ";
            std::cin >> v;
            g.printInboundEdges(v);
        }
        else if (choice == 17) { // NEW: Generate Two Random Graphs
            g.generateTwoRandomGraphs();
        }
        else if (choice == 0) {
            std::cout << "Exiting...\n";
        }
        else {
            std::cout << "Invalid choice. Try again.\n";
        }
    } while (choice != 0);

    return 0;
}

File: GRAPH THEORY/lab1/lab1_python/lab1.py
import random

class DirectedGraph:
    """A directed graph represented as two dictionaries: one for outbound edges and one for inbound edges.
    Vertices are integers from 0 to n-1. Edges are identified by (source, target) tuples with associated integer values.
    """

    def __init__(self, n):
        self._dictOut = {i: set() for i in range(n)}
        self._dictIn = {i: set() for i in range(n)}
        self._costs = {}  # dictionary with key: tuple of vertices and value: cost of the edge

    def get_number_of_vertices(self):  # Returns the total number of vertices in the graph
        return len(self._dictOut)

    def parse_vertices(self):  # Returns an iterator over all vertices in the graph
        return iter(self._dictOut.keys())

    def edge_exists(self, x, y):  # Checks if an edge from x to y exists
        return y in self._dictOut[x]

    def get_in_degree(self, vertex):  # Returns the number of incoming edges for a given vertex
        return len(self._dictIn.get(vertex, set()))

    def get_out_degree(self, vertex):  # Returns the number of outgoing edges for a given verte
        return len(self._dictOut.get(vertex, set()))

    def add_vertex(self, vertex):  # Adds a new vertex to the graph
        if vertex not in self._dictOut:
            self._dictOut[vertex] = set()
            self._dictIn[vertex] = set()

    def remove_vertex(self, vertex):  # Removes a vertex and all its associated edges from the graph
        if vertex in self._dictOut:
            for target in self._dictOut[vertex]:
                self._dictIn[target].discard(vertex)
                self._costs.pop((vertex, target), None)
            for source in self._dictIn[vertex]:
                self._dictOut[source].discard(vertex)
                self._costs.pop((source, vertex), None)
            del self._dictOut[vertex]
            del self._dictIn[vertex]

    def add_edge(self, source, target, value=0):  # Adds a directed edge from source to target with a given cost
        if source in self._dictOut and target in self._dictOut:
            if not self.edge_exists(source, target):
                self._dictOut[source].add(target)
                self._dictIn[target].add(source)
                self._costs[(source, target)] = value

    def remove_edge(self, source, target):  # Removes an edge from source to target if it exists
        if self.edge_exists(source, target):
            self._dictOut[source].discard(target)
            self._dictIn[target].discard(source)
            self._costs.pop((source, target), None)

    def get_edge_value(self, source, target):  # Retrieves the cost of an edge if it exists, otherwise returns None
        return self._costs.get((source, target), None)

    def set_edge_value(self, source, target, value):  # Updates the cost of an existing edge
        if self.edge_exists(source, target):
            self._costs[(source, target)] = value

    def get_number_of_edges(self):  # Returns the total number of edges in the graph
        return len(self._costs)

    def parse_outbound_edges(self, vertex):  # Returns an iterator over all outbound edges from a given vertex
        return iter(self._dictOut.get(vertex, set()))

    def parse_inbound_edges(self, vertex):  # Returns an iterator over all inbound edges to a given vertex
        return iter(self._dictIn.get(vertex, set()))

    def copy(self):  # Creates a deep copy of the graph
        new_graph = DirectedGraph(len(self._dictOut))
        new_graph._dictOut = {v: set(self._dictOut[v]) for v in self._dictOut}
        new_graph._dictIn = {v: set(self._dictIn[v]) for v in self._dictIn}
        new_graph._costs = self._costs.copy()
        return new_graph

    @staticmethod
    def read_from_file(filename):  # Reads graph data from a file and creates a DirectedGraph instance
        with open(filename, 'r') as f:
            n, m = map(int, f.readline().split())
            graph = DirectedGraph(n)
            for _ in range(m):
                x, y, cost = map(int, f.readline().split())
                graph.add_edge(x, y, cost)
        return graph

    def write_to_file(self, filename):  # Writes the graph data to a file
        with open(filename, 'w') as f:
            f.write(f"{len(self._dictOut)} {len(self._costs)}\n")
            for (x, y), cost in self._costs.items():
                f.write(f"{x} {y} {cost}\n")

    @staticmethod
    def random_graph(n, m):  # Generates a random directed graph with n vertices and m edges
        graph = DirectedGraph(n)
        added_edges = 0
        while added_edges < m:
            x, y = random.randint(0, n - 1), random.randint(0, n - 1)
            if x != y and not graph.edge_exists(x, y):
                graph.add_edge(x, y, random.randint(1, 100))
                added_edges += 1
        return graph


def menu():
    graph = DirectedGraph(0)

    while True:
        print("\nGraph Operations Menu:")
        print("1. Load graph from file")
        print("2. Save graph to file")
        print("3. Create a random graph")
        print("4. Display number of vertices and edges")
        print("5. Add a vertex")
        print("6. Remove a vertex")
        print("7. Add an edge")
        print("8. Remove an edge")
        print("9. Get edge cost")
        print("10. Set edge cost")
        print("11. Print the graph")
        print("12. Display outbound edges")
        print("13. Display inbound edges")
        print("14. Get in-degree of a vertex")
        print("15. Get out-degree of a vertex")
        print("16. Generate 2 random graphs (From the statement)")
        print("17. Exit")

        choice = input("Choose an option: ")

        if choice == "1":
            filename = input("Enter filename: ")
            graph = DirectedGraph.read_from_file(filename)
            print("Graph loaded successfully.")

        elif choice == "2":
            filename = input("Enter filename to save: ")
            graph.write_to_file(filename)
            print("Graph saved successfully.")

        elif choice == "3":
            n = int(input("Enter number of vertices: "))
            m = int(input("Enter number of edges: "))
            graph = DirectedGraph.random_graph(n, m)
            print("Random graph created.")

        elif choice == "4":
            print(f"Number of vertices: {graph.get_number_of_vertices()}")
            print(f"Number of edges: {graph.get_number_of_edges()}")

        elif choice == "5":
            vertex = int(input("Enter vertex to add: "))
            graph.add_vertex(vertex)
            print(f"Vertex {vertex} added.")

        elif choice == "6":
            vertex = int(input("Enter vertex to remove: "))
            graph.remove_vertex(vertex)
            print(f"Vertex {vertex} removed.")

        elif choice == "7":
            source = int(input("Enter source vertex: "))
            target = int(input("Enter target vertex: "))
            cost = int(input("Enter edge cost: "))
            graph.add_edge(source, target, cost)
            print(f"Edge {source} -> {target} with cost {cost} added.")

        elif choice == "8":
            source = int(input("Enter source vertex: "))
            target = int(input("Enter target vertex: "))
            graph.remove_edge(source, target)
            print(f"Edge {source} -> {target} removed.")

        elif choice == "9":
            source = int(input("Enter source vertex: "))
            target = int(input("Enter target vertex: "))
            cost = graph.get_edge_value(source, target)
            if cost is not None:
                print(f"Edge cost {source} -> {target}: {cost}")
            else:
                print("Edge does not exist.")

        elif choice == "10":
            source = int(input("Enter source vertex: "))
            target = int(input("Enter target vertex: "))
            cost = int(input("Enter new cost: "))
            graph.set_edge_value(source, target, cost)
            print(f"Edge {source} -> {target} updated to cost {cost}.")

        elif choice == "11":
            print("Your graph is:")
            for (x, y), cost in graph._costs.items():
                print(f"Edge {x} -> {y} with cost {cost}")

        elif choice == "12":
            vertex = int(input("Enter vertex: "))
            print(f"Outbound edges from {vertex}: {list(graph.parse_outbound_edges(vertex))}")

        elif choice == "13":
            vertex = int(input("Enter vertex: "))
            print(f"Inbound edges to {vertex}: {list(graph.parse_inbound_edges(vertex))}")

        elif choice == "14":
            vertex = int(input("Enter vertex: "))
            in_degree = graph.get_in_degree(vertex)
            print(f"In-degree of vertex {vertex}: {in_degree}")

        elif choice == "15":
            vertex = int(input("Enter vertex: "))
            out_degree = graph.get_out_degree(vertex)
            print(f"Out-degree of vertex {vertex}: {out_degree}")

        elif choice == "16":
            print(f"We are going to generate randomly 2 directed graphs with costs and save them in 2 files:")
            n1 = 7
            m1 = 20
            rgraph1 = DirectedGraph.random_graph(n1, m1)  # the statement says that is has to be a graph with 7 vertices and 20 edges
            if m1 > n1 * n1:
                raise ValueError(f"Too many edges! Max possible for {n1} vertices is {n1 * n1}.")
            else:
                rgraph1.write_to_file('random_graph1.txt')
                print(f"First graph generated and saved!")

            print(f"Let's go with the 2nd graph:")
            n2 = 6
            m2 = 40
            if m2 > n2 * n2:
                print(f"ERROR: Too many edges! Max possible for {n2} vertices is {n2 * n2}, and we have {m2} edges ")
            print("Returning to the menu...")

        elif choice == "17":
            print("Thank you, bye!.")
            break

        else:
            print("Invalid option, please try again.")


if __name__ == "__main__":
    menu()

